adapters/handlers/grpc/v1/batch/queues.go

```diff
@@ -0,0 +1,452 @@
+)
+
+// NewBatchWriteQueue creates a buffered channel to store objects for batch writing.
+//
+// The buffer size can be adjusted based on expected load and performance requirements
+// to optimize throughput and resource usage. But is required so that there is a small buffer
+// that can be quickly flushed in the event of a shutdown.
+func NewBatchInternalQueue() internalQueue {
+	return make(internalQueue, 10)
+}
+
+func NewBatchWriteQueue(buffer int) writeQueue {
+	return make(writeQueue, buffer) // Adjust buffer size as needed
+}
+
+func NewBatchWriteQueues() *WriteQueues {
+	return &WriteQueues{
+		queues: make(map[string]*WriteQueue),
+	}
+}
+
+func NewBatchReadQueues() *ReadQueues {
+	return &ReadQueues{
+		queues: make(map[string]readQueue),
+	}
+}
+
+func NewBatchReadQueue() readQueue {
+	return make(readQueue)
+}
+
+func NewStopWriteObject() *writeObject {
+	return &writeObject{
+		Object: nil,
+		Stop:   true,
+	}
+}
+
+func NewErrorsObject(errs []*pb.BatchStreamMessage_Error) *readObject {
+	return &readObject{
+		Errors: errs,
+	}
+}
+
+func NewWriteObject(obj *pb.BatchObject) *writeObject {
+	return &writeObject{
+		Object: obj,
+		Stop:   false,
+	}
+}
+
+type ReadQueues struct {
+	lock   sync.RWMutex
+	queues map[string]readQueue
+}
+
+// Get retrieves the read queue for the given stream ID.
+func (r *ReadQueues) Get(streamId string) (readQueue, bool) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	queue, ok := r.queues[streamId]
+	return queue, ok
+}
+
+// Delete removes the read queue for the given stream ID.
+func (r *ReadQueues) Delete(streamId string) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	delete(r.queues, streamId)
+}
+
+func (r *ReadQueues) Close(streamId string) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	if queue, ok := r.queues[streamId]; ok {
+		close(queue)
+	}
+}
+
+// Make initializes a read queue for the given stream ID if it does not already exist.
+func (r *ReadQueues) Make(streamId string) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	if _, ok := r.queues[streamId]; !ok {
+		r.queues[streamId] = make(readQueue)
+	}
+}
+
+type WriteQueue struct {
+	queue            writeQueue
+	consistencyLevel *pb.ConsistencyLevel
+
+	// Indexes for the next object and reference to be sent
+	objIndex int32
+	refIndex int32
+
+	lock        sync.RWMutex // Used when concurrently re-calculating NextBatchSize in Send method
+	emaQueueLen float32      // Exponential moving average of the queue length
+	buffer      int          // Buffer size for the write queue
+	alpha       float32      // Smoothing factor for EMA, typically between 0 and 1
+}
+
+// Cubic backoff function: backoff(r) = b * max(0, (r - 0.6) / 0.4) ^ 3, with b = 10s
+// E.g.
+//   - usageRatio = 0.6 -> 0s
+//   - usageRatio = 0.8 -> 1.3s
+//   - usageRatio = 0.9 -> 4.22s
+//   - usageRatio = 1.0 -> 10s
+func (w *WriteQueue) thresholdCubicBackoff(usageRatio float32) float32 {
+	b := float32(10.0) // Adjust this value as needed, defines maximum backoff in seconds
```

Comment:
nit: maybe for readability reasons we could define `maximumBackoffSeconds` const?